sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMetrics.scala

```diff
@@ -185,6 +199,75 @@ object GpuMetric extends Logging {
     }
   }
 
+  /**
+   * Give compute block, track its GpuSemaphore withholding time along with the wall time.
+   *
+   * The computation of semaphoreHoldingTime is a typical interval problem. There are 3 kinds of
+   * semaphore holding intervals:
+   *  1. The partial lead one, only the tail part is overlapped with the compute block.
+   *
+   *  2. The complete ones, which are fully overlapped with the compute block. The complete
+   *     intervals can be got from the difference of `getSemaphoreHoldingTime`, since
+   *     semaphoreHoldingTime will only be updated when the semaphore is released.
+   *
+   *  3. The partial lag one, only the head part is overlapped with the compute block.
+   *
+   * Consider the timeline and the relevant durations based on the following diagram style:
+   *
+   * Acq.     Rel.   Acq.        Rel.             Acq.          Rel.
+   *  |--------|      |-----------|                |-------------|
+   *      |----------------------------------------------| (Compute Block)
+   *      <---->      <----------->                <----->
+   *       (1)             (2)                       (3)
+   */
+  def withSemaphoreTime[T](
+      wallTime: GpuMetric,
+      semTime: GpuMetric,
+      ctx: TaskContext)(f: => T): T = {
+    require(GpuMetric.isTimeMetric(wallTime), "wallTime must be a timing metric")
+    require(GpuMetric.isTimeMetric(semTime), "semTime must be a timing metric")
+
+    val start = System.nanoTime()
+    val semStart = GpuTaskMetrics.get.getSemaphoreHoldingTime
+    // Compute the preBlockHead if there exists a partial lead interval. The preBlockHead will
+    // be explicitly subtracted in final:
+    //   Sem.Acq                                 Sem.Rel
+    //      |---------------------------------------| (Partial Lead Interval)
+    //                     |--------------------------------------| (Compute Block)
+    //        preBlockHead
+    //      <-------------->
+    val preBlockHead = GpuSemaphore.getLastSemAcqAndRelTime(ctx) match {
+      case (acqTime, relTime) if acqTime > relTime && acqTime < start =>
+        start - acqTime
+      case _ =>
+        0L
+    }
+
+    try {
+      f
+    } finally {
+      val end = System.nanoTime()
+      val semEnd = GpuTaskMetrics.get.getSemaphoreHoldingTime
+      // Compute the postBlockHead if there exists a partial lag interval. The postBlockHead will
```

Comment:
nit: strictly speaking this is not "post" block, and it's not a "head"

Reply:
Renamed